svm/src/transaction_processor.rs

```diff
@@ -494,45 +488,89 @@ impl<FG: ForkGraph> TransactionBatchProcessor<FG> {
         let fee_payer_index = 0;
         validate_fee_payer(
             fee_payer_address,
-            &mut fee_payer_account,
+            &mut loaded_fee_payer.account,
             fee_payer_index,
             error_counters,
             rent_collector,
             fee_details.total_fee(),
         )?;
 
+        if let Some(ref advanced_nonce_info) = advanced_nonce {
+            Self::validate_transaction_nonce(
+                account_loader,
+                advanced_nonce_info,
+                durable_nonce,
+                error_counters,
+            )?;
+        }
+
         // Capture fee-subtracted fee payer account and next nonce account state
         // to commit if transaction execution fails.
         let rollback_accounts = RollbackAccounts::new(
-            nonce,
+            advanced_nonce,
             *fee_payer_address,
-            fee_payer_account.clone(),
-            fee_payer_rent_debit,
+            loaded_fee_payer.account.clone(),
+            loaded_fee_payer.rent_collected,
             fee_payer_loaded_rent_epoch,
         );
 
         Ok(ValidatedTransactionDetails {
             fee_details,
             rollback_accounts,
             compute_budget_limits,
-            loaded_fee_payer_account: LoadedTransactionAccount {
-                loaded_size: fee_payer_account.data().len(),
-                account: fee_payer_account,
-                rent_collected: fee_payer_rent_debit,
-            },
+            loaded_fee_payer_account: loaded_fee_payer,
         })
     }
 
+    fn validate_transaction_nonce<CB: TransactionProcessingCallback>(
+        account_loader: &mut AccountLoader<CB>,
+        advanced_nonce_info: &NonceInfo,
+        durable_nonce: &DurableNonce,
+        error_counters: &mut TransactionErrorMetrics,
+    ) -> transaction::Result<()> {
+        // If the nonce has been used in this batch already, we must drop the transaction
+        // This is the same as if it was used in different batches in the same slot
+        // If the nonce account was closed in the batch, we error as if the blockhash didn't validate
```

Comment:
We also need to guard against the nonce account authority changing earlier in the batch by checking if the current nonce authority is passed as a signer to the advance nonce ix. Otherwise validators that split up tx batches during replay might reject a tx that uses the stale nonce authority but validators that don't split the batch would accept the tx due to the fact that the authority check was only done before starting the batch.

Reply:
good catch, youre right. i didnt realize there was even an instruction to change the nonce authority

this raises an interesting question about the opposite case tho, for the next pr where we change account locks. the goal of simd83 is to make entries an implementation detail, ie transactions in the same order should result in the same execution result regardless of how theyre batched. but what do you do with a blockhash transaction that changes a nonce authority such that the next transaction, which was previously invalid due to missing signer, becomes valid?

if you validate the batch via `check_age` as presently written, it would drop the transactions in-batch but execute it fine between batches. so you would have to defer the authority check to svm, and only check if the pseudo-blockhash matches the nonce data and that its owned by system in `check_age`. i cant think of any reasons this wouldnt be fine but it makes me nervous a bit

a safe mitigation would be if changing the nonce authority also set its data to the current durable nonce so it couldnt be used until the next slot. this would be breaking but im not sure how big the impact would be, ie if anyone actually expects to land two transactions that do this in the same slot

Reply:
> a safe mitigation would be if changing the nonce authority also set its data to the current durable nonce so it couldnt be used until the next slot. this would be breaking but im not sure how big the impact would be, ie if anyone actually expects to land two transactions that do this in the same slot

This mitigation occurred to me as well. If we made this change, then any time you update the authority on your nonce accounts you would actually completely invalidate any previously signed durable transactions because the nonce was advanced and wouldn't match the tx message hash anymore. I personally think it's ok and I was kinda surprised that this wasn't the current behavior already.

> if you validate the batch via check_age as presently written, it would drop the transactions in-batch but execute it fine between batches. so you would have to defer the authority check to svm, and only check if the pseudo-blockhash matches the nonce data and that its owned by system in check_age. i cant think of any reasons this wouldnt be fine but it makes me nervous a bit

I think this is a good approach too. I personally don't like that `check_age` does an account load currently. I would like to put everything requiring nonce data into the SVM. So the split would be:

Check (Pre-SVM):
- Message blockhash is not in recent blockhashes list
- Message starts with valid advance nonce ix (basically everything in `fn get_durable_nonce`)

SVM nonce validation:
- Load nonce and make sure it exists and is well formed
- Check if nonce is advanceable
- Ensure nonce authority signed

Reply:
>This mitigation occurred to me as well. If we made this change, then any time you update the authority on your nonce accounts you would actually completely invalidate any previously signed durable transactions because the nonce was advanced and wouldn't match the tx message hash anymore. I personally think it's ok and I was kinda surprised that this wasn't the current behavior already.

this seems arguably _more_ correct to me. the new behavior would mean all outstanding nonce transactions are invalidated. the current behavior is not so simple and, thinking about it, has some surprising edge cases

assuming keypairs A and B, and a nonce account N that starts with authority A:
* if a transaction uses N and you sign it with only B, it is invalid, but changing the authority of N from A to B makes it _become_ valid. similarly if you sign with A and change N: A -> B -> A it keeps going between valid and invalid
* if a transaction uses N and you sign it with A, then if you change the authority of N from A to B, the transaction will stay valid if B also happens to have signed it for _any_ unrelated reason

for now ill check the authority here also tho